### DISCUSSION TITLE
Nice job!But how can I use it?

### DISCUSSION THREAD

**Discussion by Cormor** [bitfireAT/dav4jvm#1] [Q&A]
.

**Comment by rossbeazley** [comment]
did you get anywhere?

i have authentication but getting 400 trying to list a directory

```
val httpClient = OkHttpClient.Builder()
            .followRedirects(false)
            .authenticator(object : Authenticator {
                override fun authenticate(route: Route?, response: okhttp3.Response): Request {
                    return response.request.newBuilder()
                        .header("Authorization", Credentials.basic(
                            username = "",
                            password = ""
                        ))
                        .build()
                }

            })
            .build()

        val davCollection = DavCollection(
            httpClient,
            "https://urlz/".toHttpUrl()
        )

        davCollection.propfind(1) { a: Response, b: Response.HrefRelation ->
            println(a)
        }
```

**Comment by rfc2822** [comment]
Looks correct. You'd have to debug the HTTP request itself by adding a logger interceptor. See the DAVx5 source code: https://github.com/bitfireAT/davx5-ose/

**Comment by rossbeazley** [comment]
thanks, i was also inspired to actually check the server logs (who would have thought! :) ) so i have a thread to pull now as it reports

{"method": "PROPFIND", "path": "/dav/personal-files/webdavdev/", "error": "prop must not be empty"}

**Comment by rossbeazley** [comment]
and now it works with 

        davCollection.propfind(1, DisplayName.NAME) { a: Response, b: Response.HrefRelation ->
            println(a)
        }
        
thanks for the pointers

**Reply by rfc2822** [reply]
Yes that means that you're interested in the `DisplayName` and the server should return it.

**Comment by rossbeazley** [comment]
My use case is photo upload to my photoprism instance (https://photoprism.app/) from my android phone.

Would you be interested in a how to or similar document once I have my solution working, i expect things I need to do include;
getting file lists from remote,
maybe ordered by the most recently created first,
file upload,
directory creation,
authentication

It might help folks getting started

**Reply by rfc2822** [reply]
We have written the library primarily for DAVx⁵, but it's open-source and the intention is that people use if it is useful for them :)

Unfortunately, our time resources are very limited. If you're interested in writing a short tutorial (maybe just the most important calls that you have already used) in the repo's README.md, please do so in a fork and send a pull request, I'll have a look at it and merge it.

**Comment by rossbeazley** [comment]
i see now i can use BasicDigestAuthHandler(....) 

```
val httpClient = OkHttpClient.Builder()
            .followRedirects(false)
            .authenticator(BasicDigestAuthHandler(
                domain = null,
                username = "",
                password = "")
            )
            .build()
```
so much simpler
